File: frontend/src/api/github.ts
import apiClient, { ANALYSIS_TIMEOUT } from './client'

export interface GitHubRepo {
  id: number
  name: string
  full_name: string
  description: string | null
  html_url: string
  clone_url: string
  language: string | null
  stargazers_count: number
  forks_count: number
  created_at: string
  updated_at: string
  pushed_at: string | null
  fork: boolean
  owner: string
}

// LLM-generated content types (v1.2)
export interface ImplementationDetail {
  title: string
  items: string[]
}

export interface DevelopmentTimelinePhase {
  period: string
  title: string
  activities: string[]
}

export interface DetailedAchievementItem {
  title: string
  description: string
}

export interface RepoAnalysis {
  id: number
  project_id: number
  git_url: string
  total_commits: number
  user_commits: number
  lines_added: number
  lines_deleted: number
  files_changed: number
  languages: Record<string, number>
  primary_language: string | null
  detected_technologies: string[]
  commit_messages_summary: string | null
  commit_categories: Record<string, number> | null
  architecture_patterns: string[] | null
  key_tasks: string[] | null  // LLM-generated key tasks
  // LLM-generated detailed content (v1.2)
  implementation_details: ImplementationDetail[] | null
  development_timeline: DevelopmentTimelinePhase[] | null
  tech_stack_versions: Record<string, string[]> | null
  detailed_achievements: Record<string, DetailedAchievementItem[]> | null
  // AI-generated summary (v1.12 - generated during analysis)
  ai_summary?: string | null
  ai_key_features?: string[] | null
  // User's code contributions context (v1.12)
  user_code_contributions?: {
    summary?: {
      total_commits?: number
      analyzed_commits?: number
      lines_added?: number
      lines_deleted?: number
      files_changed?: number
    }
    technologies?: string[]
    work_areas?: string[]
  } | null
  analyzed_at: string
  // LLM usage tracking (v1.8)
  provider?: string | null
  token_usage?: number | null
  // Auto-calculated contribution percent (v1.12)
  suggested_contribution_percent?: number | null
  // Language used for analysis (v1.12)
  analysis_language?: string  // "ko" or "en"
}

export interface RepoQuickInfo {
  name: string
  description: string | null
  html_url: string
  start_date: string | null
  end_date: string | null
  team_size: number
  contribution_percent: number
  total_commits: number
  user_commits: number
}

export interface AIGeneratedDescription {
  short_description: string | null
  description: string | null
  technologies: string[]
}

export interface GitHubStatus {
  connected: boolean
  github_username: string | null
  avatar_url: string | null
  valid: boolean
  message?: string
}

export interface FileTreeItem {
  type: 'file' | 'directory'
  name: string
  path: string
  size: number
  sha: string
  download_url?: string
}

export interface FileTreeResponse {
  files: FileTreeItem[]
  path: string
  git_url: string
}

export interface FileContentResponse {
  name: string
  path: string
  size: number
  sha: string
  content: string
  encoding: string
  download_url: string | null
}

// Batch import types
export interface ImportRepoResult {
  repo_url: string
  project_id?: number
  project_name: string
  success: boolean
  message: string
}

export interface ImportReposResponse {
  imported: number
  failed: number
  results: ImportRepoResult[]
}

export interface BatchAnalysisResult {
  project_id: number
  project_name: string
  success: boolean
  message: string
  detected_technologies?: string[]
  detected_role?: string
}

export interface BatchAnalysisResponse {
  task_id?: string
  total: number
  completed: number
  failed: number
  results: BatchAnalysisResult[]
}

// Inline editing types
export interface EditStatus {
  key_tasks_modified: boolean
  implementation_details_modified: boolean
  detailed_achievements_modified: boolean
}

export interface EffectiveRepoAnalysis extends RepoAnalysis {
  edit_status: EditStatus
}

export interface AnalysisContentUpdate {
  field: 'key_tasks' | 'implementation_details' | 'detailed_achievements'
  content: any
}

export interface AnalysisUpdateResponse {
  success: boolean
  field: string
  message: string
}

// ============ Extended Analysis Types (v1.10) ============

export interface DetailedCommit {
  sha: string
  short_sha: string
  message: string
  full_message?: string
  author: string
  author_email?: string
  date: string
  commit_type: string  // feat, fix, refactor, docs, test, chore, perf, style, other
  type_label: string  // Human-readable label
  scope?: string
  description: string
  is_breaking: boolean
  files_changed: number
  lines_added: number
  lines_deleted: number
  work_areas: string[]  // ["frontend", "backend", "tests", ...]
}

export interface ContributorSummary {
  username: string
  avatar_url?: string
  contributions: number
  html_url?: string
}

export interface ContributorAnalysis {
  username: string
  email?: string
  is_primary: boolean
  total_commits: number
  first_commit_date?: string
  last_commit_date?: string
  lines_added: number
  lines_deleted: number
  file_extensions: Record<string, number>  // {".py": 45, ".ts": 30}
  work_areas: string[]  // ["frontend", "backend", "tests"]
  detected_technologies: string[]
  detailed_commits: DetailedCommit[]
  commit_types: Record<string, number>  // {"feat": 40, "fix": 30}
}

export interface FileCountByType {
  code: number
  test: number
  docs: number
  config: number
}

export interface CodeQualityMetrics {
  total_files: number
  total_lines: number  // Estimated
  avg_file_size: number
  max_file_size: number
  test_file_ratio: number  // Percentage of test files
  doc_file_ratio: number  // Percentage of doc files
  code_file_ratio: number  // Percentage of code files
  config_file_count: number
  language_distribution: Record<string, number>  // {".py": 45.5, ".ts": 30.2}
  file_count_by_type: FileCountByType
}

export interface ContributorsListResponse {
  contributors: ContributorSummary[]
  total: number
}

export interface DetailedCommitsResponse {
  commits: DetailedCommit[]
  total: number
  author?: string
}

export interface ExtendedRepoAnalysis extends RepoAnalysis {
  repo_technologies: string[]
  all_contributors: ContributorSummary[]
  code_quality_metrics?: CodeQualityMetrics
  contributors: ContributorAnalysis[]
  primary_contributor?: ContributorAnalysis
}

// ============ Background Analysis Types (v1.12) ============

export interface AnalysisJobStatus {
  task_id: string
  project_id: number
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled'
  progress: number
  current_step: number
  total_steps: number
  step_name?: string
  error_message?: string
  partial_results?: Record<string, any>
  started_at?: string
  completed_at?: string
  created_at: string
  token_usage?: number  // Total tokens used in LLM calls
  llm_provider?: string  // LLM provider used (e.g., 'openai', 'anthropic', 'claude_code', 'gemini_cli')
}

export interface AnalysisJobListResponse {
  jobs: AnalysisJobStatus[]
  total: number
}

export interface StartAnalysisResponse {
  task_id: string
  project_id: number
  message: string
}

export interface CancelAnalysisResponse {
  task_id: string
  project_id: number
  status: string
  message: string
  partial_saved: boolean
}

export const githubApi = {
  getStatus: (userId: number) =>
    apiClient.get<GitHubStatus>('/github/status', {
      params: { user_id: userId }
    }),

  connect: (redirectUrl?: string, isElectron?: boolean, userId?: number) =>
    apiClient.get<{ auth_url: string }>('/github/connect', {
      params: {
        redirect_url: redirectUrl,
        frontend_origin: window.location.origin,  // Pass current origin for dynamic redirect
        is_electron: isElectron || false,  // Use custom protocol for Electron OAuth callback
        user_id: userId  // Link GitHub to existing user instead of creating new one
      }
    }),

  disconnect: (userId: number) =>
    apiClient.delete('/github/disconnect', { params: { user_id: userId } }),

  getRepos: (userId: number, fetchAll: boolean = true) =>
    apiClient.get<{ repos: GitHubRepo[]; total: number; has_more: boolean }>('/github/repos', {
      params: { user_id: userId, fetch_all: fetchAll }
    }),

  analyzeRepo: (
    userId: number,
    gitUrl: string,
    projectId?: number,
    options?: {
      provider?: string
      cli_mode?: 'claude_code' | 'gemini_cli'
      cli_model?: string
      language?: 'ko' | 'en'  // Analysis language (v1.12)
    }
  ) =>
    apiClient.post<RepoAnalysis>('/github/analyze', {
      git_url: gitUrl,
      project_id: projectId
    }, {
      timeout: ANALYSIS_TIMEOUT,  // Extended timeout for CLI/LLM analysis
      params: {
        user_id: userId,
        provider: options?.provider,
        cli_mode: options?.cli_mode,
        cli_model: options?.cli_model,
        language: options?.language  // Pass analysis language
      }
    }),

  getAnalysis: (projectId: number) =>
    apiClient.get<RepoAnalysis>(`/github/analysis/${projectId}`),

  getRepoInfo: (userId: number, gitUrl: string) =>
    apiClient.get<RepoQuickInfo>('/github/repo-info', {
      params: { user_id: userId, git_url: gitUrl }
    }),

  generateDescription: (userId: number, gitUrl: string) =>
    apiClient.post<AIGeneratedDescription>('/github/generate-description', {
      git_url: gitUrl
    }, { params: { user_id: userId } }),

  detectTechnologies: (userId: number, gitUrl: string) =>
    apiClient.get<{ technologies: string[] }>('/github/detect-technologies', {
      params: { user_id: userId, git_url: gitUrl }
    }),

  getFileTree: (userId: number, gitUrl: string, options?: {
    path?: string
    ref?: string
    recursive?: boolean
  }) =>
    apiClient.get<FileTreeResponse>('/github/file-tree', {
      params: {
        user_id: userId,
        git_url: gitUrl,
        path: options?.path || '',
        ref: options?.ref,
        recursive: options?.recursive || false
      }
    }),

  getFileContent: (userId: number, gitUrl: string, filePath: string, ref?: string) =>
    apiClient.get<FileContentResponse>('/github/file-content', {
      params: {
        user_id: userId,
        git_url: gitUrl,
        file_path: filePath,
        ref
      }
    }),

  // Batch operations
  importRepos: (userId: number, repoUrls: string[], autoAnalyze: boolean = false) =>
    apiClient.post<ImportReposResponse>('/github/import-repos', {
      repo_urls: repoUrls,
      auto_analyze: autoAnalyze
    }, { params: { user_id: userId } }),

  analyzeBatch: (
    userId: number,
    projectIds: number[],
    options?: {
      llm_provider?: string
      cli_mode?: 'claude_code' | 'gemini_cli'
      cli_model?: string
    }
  ) =>
    apiClient.post<BatchAnalysisResponse>('/github/analyze-batch', {
      project_ids: projectIds,
      llm_provider: options?.llm_provider,
      cli_mode: options?.cli_mode,
      cli_model: options?.cli_model
    }, {
      timeout: ANALYSIS_TIMEOUT * projectIds.length,  // Extended timeout based on project count
      params: { user_id: userId }
    }),

  // Inline editing APIs
  getEffectiveAnalysis: (projectId: number) =>
    apiClient.get<EffectiveRepoAnalysis>(`/github/analysis/${projectId}/effective`),

  updateAnalysisContent: (projectId: number, update: AnalysisContentUpdate) =>
    apiClient.patch<AnalysisUpdateResponse>(`/github/analysis/${projectId}/content`, update),

  resetAnalysisField: (projectId: number, field: string) =>
    apiClient.post<AnalysisUpdateResponse>(`/github/analysis/${projectId}/reset/${field}`),

  // Extended analysis APIs (v1.10)
  getContributors: (projectId: number, userId: number) =>
    apiClient.get<ContributorsListResponse>(`/github/contributors/${projectId}`, {
      params: { user_id: userId }
    }),

  getContributorAnalysis: (
    projectId: number,
    userId: number,
    options?: {
      username?: string
      refresh?: boolean
    }
  ) =>
    apiClient.get<ContributorAnalysis>(`/github/contributor-analysis/${projectId}`, {
      params: {
        user_id: userId,
        username: options?.username,
        refresh: options?.refresh || false
      },
      timeout: ANALYSIS_TIMEOUT  // Extended timeout for fresh analysis
    }),

  getCodeQuality: (projectId: number, userId: number) =>
    apiClient.get<CodeQualityMetrics>(`/github/code-quality/${projectId}`, {
      params: { user_id: userId }
    }),

  getDetailedCommits: (
    projectId: number,
    userId: number,
    options?: {
      author?: string
      limit?: number
    }
  ) =>
    apiClient.get<DetailedCommitsResponse>(`/github/detailed-commits/${projectId}`, {
      params: {
        user_id: userId,
        author: options?.author,
        limit: options?.limit || 50
      }
    }),

  // ============ Background Analysis APIs (v1.12) ============

  startBackgroundAnalysis: (
    userId: number,
    gitUrl: string,
    projectId?: number,
    options?: {
      provider?: string
      cli_mode?: 'claude_code' | 'gemini_cli'
      cli_model?: string
      language?: 'ko' | 'en'  // Analysis language (v1.12)
    }
  ) =>
    apiClient.post<StartAnalysisResponse>('/github/analyze-background', {
      git_url: gitUrl,
      project_id: projectId
    }, {
      params: {
        user_id: userId,
        provider: options?.provider,
        cli_mode: options?.cli_mode,
        cli_model: options?.cli_model,
        language: options?.language  // Pass analysis language
      }
    }),

  getActiveAnalyses: (userId: number) =>
    apiClient.get<AnalysisJobListResponse>('/github/active-analyses', {
      params: { user_id: userId }
    }),

  getAnalysisStatus: (projectId: number, userId: number) =>
    apiClient.get<AnalysisJobStatus | null>(`/github/analysis-status/${projectId}`, {
      params: { user_id: userId }
    }),

  cancelAnalysis: (projectId: number, userId: number) =>
    apiClient.post<CancelAnalysisResponse>(`/github/analysis/${projectId}/cancel`, null, {
      params: { user_id: userId }
    }),

  getJobStatus: (taskId: string, userId: number) =>
    apiClient.get<AnalysisJobStatus>(`/github/job/${taskId}`, {
      params: { user_id: userId }
    }),

  // Save GitHub token (used by Electron desktop app via gh CLI)
  saveToken: (userId: number, token: string) =>
    apiClient.post<{
      success: boolean
      message: string
      user_id: number  // The actual user_id where token was saved (may differ due to merge)
      github_username?: string
      github_avatar_url?: string
      merged_from_user_id?: number  // If token was merged to existing user
    }>('/github/save-token', null, {
      params: { user_id: userId, token }
    }),
}

File: api/schemas/github.py
from pydantic import BaseModel
from datetime import datetime
from typing import Optional, List, Dict, Any


class GitHubConnectRequest(BaseModel):
    """Request to initiate GitHub OAuth flow."""

    redirect_url: Optional[str] = None


class GitHubCallbackResponse(BaseModel):
    """Response after GitHub OAuth callback."""

    success: bool
    user_id: int
    github_username: str
    github_avatar_url: Optional[str] = None
    message: str


class GitHubUserInfo(BaseModel):
    """GitHub user information."""

    login: str
    id: int
    avatar_url: Optional[str] = None
    name: Optional[str] = None
    email: Optional[str] = None
    public_repos: int = 0
    followers: int = 0
    following: int = 0


class GitHubRepoInfo(BaseModel):
    """GitHub repository information."""

    id: int
    name: str
    full_name: str
    description: Optional[str] = None
    html_url: str
    clone_url: str
    language: Optional[str] = None
    stargazers_count: int = 0
    forks_count: int = 0
    created_at: datetime
    updated_at: datetime
    pushed_at: Optional[datetime] = None
    fork: bool = False
    owner: str = ""


class RepoAnalysisRequest(BaseModel):
    """Request to analyze a GitHub repository."""

    git_url: str
    project_id: Optional[int] = None  # Associate with existing project
    summary_style: Optional[str] = (
        None  # "professional", "casual", "technical" - uses user default if not specified
    )


class RepoAnalysisResponse(BaseModel):
    """Response from repository analysis."""

    id: int
    project_id: int
    git_url: str
    total_commits: int
    user_commits: int
    lines_added: int
    lines_deleted: int
    files_changed: int
    languages: Dict[str, float]
    primary_language: Optional[str] = None
    detected_technologies: List[str]
    commit_messages_summary: Optional[str] = None
    commit_categories: Optional[Dict[str, int]] = None
    architecture_patterns: Optional[List[str]] = None
    key_tasks: Optional[List[str]] = None  # LLM-generated key tasks
    # LLM-generated detailed content (v1.2)
    implementation_details: Optional[List[Dict[str, Any]]] = None  # [{title, items}]
    development_timeline: Optional[List[Dict[str, Any]]] = (
        None  # [{period, title, activities}]
    )
    tech_stack_versions: Optional[Dict[str, List[str]]] = (
        None  # {Frontend: [...], Backend: [...]}
    )
    detailed_achievements: Optional[Dict[str, List[Dict[str, str]]]] = (
        None  # {category: [{title, description}]}
    )
    # AI-generated summary (v1.12)
    ai_summary: Optional[str] = None
    ai_key_features: Optional[List[str]] = None
    # User's code contributions context (v1.12)
    user_code_contributions: Optional[Dict[str, Any]] = None
    analyzed_at: datetime
    # LLM usage tracking (v1.8)
    provider: Optional[str] = None
    token_usage: Optional[int] = None
    # Auto-calculated contribution percent (v1.12)
    suggested_contribution_percent: Optional[int] = None
    # Language used for analysis (v1.12)
    analysis_language: str = "ko"  # "ko" or "en"
    # AI tools detected from commits (v1.17)
    ai_tools_detected: Optional[List[Dict[str, Any]]] = None

    class Config:
        from_attributes = True


class GitHubRepoListResponse(BaseModel):
    """List of user's GitHub repositories."""

    repos: List[GitHubRepoInfo]
    total: int
    page: int = 1
    per_page: int = 100
    has_more: bool = False
    cached: bool = False
    cached_at: Optional[str] = None


class ImportReposRequest(BaseModel):
    """Request to import multiple repos as projects."""

    repo_urls: List[str]
    auto_analyze: bool = False


class ImportRepoResult(BaseModel):
    """Result of importing a single repo."""

    repo_url: str
    project_id: Optional[int] = None
    project_name: str
    success: bool
    message: str


class ImportReposResponse(BaseModel):
    """Response from batch import."""

    imported: int
    failed: int
    results: List[ImportRepoResult]


class BatchAnalysisRequest(BaseModel):
    """Request to analyze multiple projects."""

    project_ids: List[int]
    # LLM settings (v1.8)
    llm_provider: Optional[str] = None  # 'openai', 'anthropic', 'gemini'
    cli_mode: Optional[str] = None  # 'claude_code' or 'gemini_cli'
    cli_model: Optional[str] = None  # Model name for CLI


class BatchAnalysisResult(BaseModel):
    """Result of analyzing a single project."""

    project_id: int
    project_name: str
    success: bool
    message: str
    detected_technologies: Optional[List[str]] = None
    detected_role: Optional[str] = None


class BatchAnalysisResponse(BaseModel):
    """Response from batch analysis."""

    task_id: Optional[str] = None
    total: int
    completed: int
    failed: int
    results: List[BatchAnalysisResult]


# ============ Inline Editing Schemas ============


class AnalysisContentUpdate(BaseModel):
    """Request to update analysis content."""

    field: str  # 'key_tasks', 'implementation_details', or 'detailed_achievements'
    content: Any  # The updated content (type depends on field)


class ImplementationDetailItem(BaseModel):
    """Single implementation detail item."""

    title: str
    items: List[str]


class DetailedAchievementItem(BaseModel):
    """Single detailed achievement item."""

    title: str
    description: str


class EditStatus(BaseModel):
    """Status of edited fields."""

    key_tasks_modified: bool = False
    implementation_details_modified: bool = False
    detailed_achievements_modified: bool = False


class EffectiveAnalysisResponse(BaseModel):
    """Analysis response with user edits applied."""

    id: int
    project_id: int
    git_url: str
    total_commits: int
    user_commits: int
    lines_added: int
    lines_deleted: int
    files_changed: int
    languages: Dict[str, float]
    primary_language: Optional[str] = None
    detected_technologies: List[str]
    commit_messages_summary: Optional[str] = None
    commit_categories: Optional[Dict[str, int]] = None
    architecture_patterns: Optional[List[str]] = None
    # These fields may contain edited content
    key_tasks: Optional[List[str]] = None
    implementation_details: Optional[List[Dict[str, Any]]] = None
    development_timeline: Optional[List[Dict[str, Any]]] = None
    tech_stack_versions: Optional[Dict[str, List[str]]] = None
    detailed_achievements: Optional[Dict[str, List[Dict[str, str]]]] = None
    # AI-generated summary (v1.12)
    ai_summary: Optional[str] = None
    ai_key_features: Optional[List[str]] = None
    analyzed_at: datetime
    # Edit status
    edit_status: EditStatus
    # Auto-calculated contribution percent (v1.12)
    suggested_contribution_percent: Optional[int] = None
    # Language used for analysis (v1.12)
    analysis_language: str = "ko"  # "ko" or "en"
    # AI tools detected from commits (v1.17)
    ai_tools_detected: Optional[List[Dict[str, Any]]] = None

    class Config:
        from_attributes = True


class RepoAnalysisSummary(BaseModel):
    """Per-repo analysis summary for multi-repo projects."""

    repo_url: str
    label: Optional[str] = None
    is_primary: bool = False
    total_commits: int = 0
    user_commits: int = 0
    lines_added: int = 0
    lines_deleted: int = 0
    files_changed: int = 0
    detected_technologies: List[str] = []
    primary_language: Optional[str] = None
    key_tasks: Optional[List[str]] = None
    # Rich per-repo fields (v1.17)
    implementation_details: Optional[List[Dict[str, Any]]] = None
    detailed_achievements: Optional[Dict[str, List[Dict[str, str]]]] = None
    ai_summary: Optional[str] = None
    ai_key_features: Optional[List[str]] = None
    languages: Optional[Dict[str, float]] = None
    commit_categories: Optional[Dict[str, int]] = None
    development_timeline: Optional[List[Dict[str, Any]]] = None
    ai_tools_detected: Optional[List[Dict[str, Any]]] = None

    class Config:
        from_attributes = True


class MultiRepoAnalysisResponse(BaseModel):
    """All per-repo analyses for a multi-repo project."""

    project_id: int
    repo_count: int
    analyses: List[RepoAnalysisSummary]

    class Config:
        from_attributes = True


# ============ Extended Analysis Schemas (v1.10) ============


class DetailedCommit(BaseModel):
    """Detailed commit information with Conventional Commit parsing."""

    sha: str
    short_sha: Optional[str] = None  # Will be derived from sha if not provided
    message: str
    full_message: Optional[str] = None
    author: str
    author_email: Optional[str] = None
    date: datetime
    commit_type: str = (
        "other"  # feat, fix, refactor, docs, test, chore, perf, style, other
    )
    type_label: str = "Other"  # Human-readable label
    scope: Optional[str] = None
    description: Optional[str] = None  # Will be derived from message if not provided
    is_breaking: bool = False
    files_changed: int = 0
    lines_added: int = 0
    lines_deleted: int = 0
    work_areas: List[str] = []  # ["frontend", "backend", "tests", ...]

    def model_post_init(self, __context) -> None:
        """Fill in derived fields after model initialization."""
        if self.short_sha is None and self.sha:
            object.__setattr__(self, "short_sha", self.sha[:7])
        if self.description is None and self.message:
            object.__setattr__(self, "description", self.message.split("\n")[0][:100])


class ContributorSummary(BaseModel):
    """Summary info for a single contributor."""

    username: str
    avatar_url: Optional[str] = None
    contributions: int = 0
    html_url: Optional[str] = None


class ContributorAnalysisResponse(BaseModel):
    """Full contributor analysis response."""

    username: str
    email: Optional[str] = None
    is_primary: bool = False
    total_commits: int = 0
    first_commit_date: Optional[datetime] = None
    last_commit_date: Optional[datetime] = None
    lines_added: int = 0
    lines_deleted: int = 0
    file_extensions: Dict[str, int] = {}  # {".py": 45, ".ts": 30}
    work_areas: List[str] = []  # ["frontend", "backend", "tests"]
    detected_technologies: List[str] = []
    detailed_commits: List[DetailedCommit] = []
    commit_types: Dict[str, int] = {}  # {"feat": 40, "fix": 30}

    class Config:
        from_attributes = True


class FileCountByType(BaseModel):
    """File counts by type."""

    code: int = 0
    test: int = 0
    docs: int = 0
    config: int = 0


class CodeQualityMetrics(BaseModel):
    """Code quality metrics for a repository."""

    total_files: int = 0
    total_lines: int = 0  # Estimated
    avg_file_size: float = 0
    max_file_size: int = 0
    test_file_ratio: float = 0  # Percentage of test files
    doc_file_ratio: float = 0  # Percentage of doc files
    code_file_ratio: float = 0  # Percentage of code files
    config_file_count: int = 0
    language_distribution: Dict[str, float] = {}  # {".py": 45.5, ".ts": 30.2}
    file_count_by_type: FileCountByType = FileCountByType()


class ContributorsListResponse(BaseModel):
    """List of all contributors in a repository."""

    contributors: List[ContributorSummary]
    total: int


class ExtendedRepoAnalysisResponse(RepoAnalysisResponse):
    """Extended analysis response with contributor and quality metrics."""

    # Repository-level technologies (separate from user-detected)
    repo_technologies: List[str] = []
    # All contributors summary
    all_contributors: List[ContributorSummary] = []
    # Code quality metrics
    code_quality_metrics: Optional[CodeQualityMetrics] = None
    # Contributor analyses
    contributors: List[ContributorAnalysisResponse] = []
    # Primary contributor (logged-in user)
    primary_contributor: Optional[ContributorAnalysisResponse] = None


# ============ Analysis Job Schemas (v1.12) ============


class AnalysisJobStatus(BaseModel):
    """Status of an analysis job."""

    task_id: str
    project_id: Optional[int] = None
    status: str  # pending, running, completed, failed, cancelled
    progress: int  # 0-100
    current_step: int
    total_steps: int
    step_name: Optional[str] = None
    error_message: Optional[str] = None
    partial_results: Optional[Dict[str, Any]] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    token_usage: Optional[int] = None  # Total tokens used in LLM calls
    llm_provider: Optional[str] = None  # LLM provider used

    class Config:
        from_attributes = True


class AnalysisJobListResponse(BaseModel):
    """List of active analysis jobs."""

    jobs: List[AnalysisJobStatus]
    total: int


class StartAnalysisResponse(BaseModel):
    """Response when starting a background analysis."""

    task_id: str
    project_id: int
    message: str


class CancelAnalysisResponse(BaseModel):
    """Response when cancelling an analysis."""

    task_id: str
    project_id: int
    status: str
    message: str
    partial_saved: bool = False
